internal/controller/controllers/events_uploader_handler.go

```diff
@@ -0,0 +1,304 @@
+package controllers
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"strings"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/openshift/assisted-service/internal/common"
+	eventsapi "github.com/openshift/assisted-service/internal/events/api"
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/pkg/k8sclient"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type serializedAuthMap struct {
+	Auths map[string]serializedAuth `json:"auths"`
+}
+type serializedAuth struct {
+	Auth string `json:"auth"`
+}
+
+type EventsUploaderConfig struct {
+	DataUploadEndpoint     string `envconfig:"DATA_UPLOAD_ENDPOINT" default:""`
+	AssistedServiceVersion string
+	EnableDataCollection   bool `envconfig:"ENABLE_DATA_COLLECTION" default:"true"`
+}
+
+type uploader struct {
+	EventsUploaderConfig
+	db     *gorm.DB
+	log    log.FieldLogger
+	client k8sclient.K8SClient
+}
+
+const (
+	openshiftTokenKey = "cloud.openshift.com"
+)
+
+func NewEventsUploaderWrapper(cfg *EventsUploaderConfig, db *gorm.DB, log logrus.FieldLogger, client k8sclient.K8SClient) *uploader {
+	return &uploader{
+		db:                   db,
+		log:                  log,
+		client:               client,
+		EventsUploaderConfig: *cfg,
+	}
+}
+
+func (u *uploader) UploadEvents(ctx context.Context, clusterID *strfmt.UUID, eventsHandler eventsapi.Handler) error {
+	if u.EnableDataCollection {
```

Comment:
How about having a dummy uploader in main that will be created if the feature is not enabled?

Reply:
I think this could be a good reason to use interfaces for this.
I imagine it would be similar to how we use the `NoneAuthenticator`.

Reply:
Good points! Created a dummy uploader [here](https://github.com/openshift/assisted-service/pull/4880/files#diff-ee4f3756eb0ee07d93469c4f3c7182fbaf5549c1e51f700ea4b4b32dddfdfbfdR1) and [`NewUploadHandler`](https://github.com/openshift/assisted-service/pull/4880/files#diff-97008778ca8858564f55e9743e204f1f2a4a6475a2fe58517fdbff858d4e8e5aR23) will create either the actual event uploader or the dummy uploader depending on if it's enabled.